### PR TITLE
Bios third person validation

### PR DESCRIPTION
**Pull request by RossellaFer**
#119 - custom validation for I and わたし

**Proposed Changes**
Added vuelidate custom validation to find out whether the string in the bio field contains the words I, わたし or 私

*Description of changes in this PR*
New custom validation in its own `.js` file and added validations to the NominateForm. Added unit tests with some sample sentences.
*Japanese translation of the error message missing*

**Testing Plan**
Copy and paste the following examples to verify that the error message appears only when  I, わたし or 私 are in the bio. 

```
Should error:
I'm a full-stack developer with a background in Japanese culture.
I work as a full stack developer for one of the best companies in Japan.

Should pass:
D&I Lead at Hogwarts school of witchcraft and wizardry
She is a volunteer for Women Who Code Tokyo

```

### CODE REVIEW

**Comment by ann-kilzer** [comment]
Thanks for making this PR! An interesting case to consider: Roman numerals, e.g “Title I “ Or perhaps this is too rare? I wonder if there is a way to make this a gentle reminder rather than an all or nothing? 

Also Japanese has other first person pronouns besides 私 like ぼく and あたし. Read more here: https://en.wikipedia.org/wiki/Japanese_pronouns

**Comment by ann-kilzer** [comment]
Another thought, we might consider the subject marking participle “は” to distinguish from other words that could possibly overlap with わたし or あたし, ぼく... would be great to get Ania to review this as well.

File: web/tests/unit/validator-third-person.spec.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  });
+  it('accepts "I" if not first person', () => {
+    expect(isThirdPerson('She works in D&I')).to.be.true;
+  });
+  it('rejects first person in English', () => {
+    expect(isThirdPerson('I\' m a front end developer')).to.be.false;
+    expect(isThirdPerson('While in London, I studied things.')).to.be.false;
+  });
+  it('rejects first person in Japanese', () => {
+    expect(isThirdPerson('私わイタリア人です。')).to.be.false;
```
Comment (ann-kilzer) [inline web/tests/unit/validator-third-person.spec.js:20]: nit: わ => は

File: web/tests/unit/validator-third-person.spec.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/* eslint-disable no-unused-expressions */
```
Comment (ann-kilzer) [inline web/tests/unit/validator-third-person.spec.js:1]: Thanks for writing tests!

**Review by ann-kilzer — COMMENTED** [review]
First pass

**Comment by ann-kilzer** [comment]
Some other test cases to consider include words ending with 私. https://jisho.org/word/%E5%85%AC%E7%A7%81

**Comment by Ania-chan** [comment]
@ann-kilzer @RossellaFer I went through all the introductions written in Japanese and all of them are in 3rd person. Generally in Japanese in these kind of introductions pronouns are very very rarely used. Even though in Japanese there are many different pronouns, in this kind of formal writing only 私 is used (わたし in hiragana isn't used unless it's a mistake).
I think there's a possibility that some no native Japanese speakers might add their description in Japanese so in this case we might want to check for 私 cause it's a common mistake to use it but I don't think that we need this check if we only consider Japanese speakers.

**Comment by ann-kilzer** [comment]
Thanks for the perspective! It will keep it simpler too.

File: web/tests/unit/validator-third-person.spec.js (hunk 1 above)
Comment (RossellaFer) [reply web/tests/unit/validator-third-person.spec.js]: Just corrected this, thanks for checking!

File: web/src/validators/isThirdPerson.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// custom vuelidate validator to ensure input does not contain first person
+const isThirdPerson = (value) => {
```
Comment (ann-kilzer) [inline web/src/validators/isThirdPerson.js:2]: The naming is confusing from a grammar perspective (which includes 1st, 2nd, and 3rd, singular and plural in English), I would just call this isFirstPerson and negate it later on. The file will also need renaming, and the tests 😑 ... sorry it's a lot of flipping.
```suggestion
const isFirstPerson = (value) => {
```

File: web/src/validators/isThirdPerson.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+  const firstPersonRegex = /((?<=[\s,.:;"’']|^)I(?=[\s,.:;"’'])|わたし|私)/;
```
Comment (ann-kilzer) [inline web/src/validators/isThirdPerson.js:3]: Adding subject marking participle 'は' and `^`. Thoughts?
```suggestion
  const firstPersonRegex = /((?<=[\s,.:;"’']|^)I(?=[\s,.:;"’'])|^わたしは|^私は)/;
```

File: web/src/validators/isThirdPerson.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+
+  return !firstPersonRegex.test(value);
+};
+
+export default isThirdPerson;
```
Comment (ann-kilzer) [inline web/src/validators/isThirdPerson.js:8]:
```suggestion
export default isFirstPerson;
```

File: web/src/validators/isThirdPerson.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// custom vuelidate validator to ensure input does not contain first person
+const isThirdPerson = (value) => {
+  const firstPersonRegex = /((?<=[\s,.:;"’']|^)I(?=[\s,.:;"’'])|わたし|私)/;
+
+  return !firstPersonRegex.test(value);
```
Comment (ann-kilzer) [inline web/src/validators/isThirdPerson.js:5]:
```suggestion
  return firstPersonRegex.test(value);
```

File: web/src/components/nomination/NominateForm.vue [hunk 7]
```diff
@@ -133,6 +133,7 @@ import {
   required, email, minLength, url,
 } from 'vuelidate/lib/validators';
 import japanese from '@/validators/japanese';
+import isThirdPerson from '@/validators/isThirdPerson';
```
Comment (ann-kilzer) [inline web/src/components/nomination/NominateForm.vue:136]:
```suggestion
import isFirstPerson from '@/validators/isFirstPerson';
```

File: web/src/components/nomination/NominateForm.vue [hunk 8]
```diff
@@ -166,6 +167,7 @@ export default {
       speaker_bio: {
         required,
         minLength: minLength(BIO_LENGTH),
+        isThirdPerson,
```
Comment (ann-kilzer) [inline web/src/components/nomination/NominateForm.vue:170]:
```suggestion
        isFirstPerson,
```

File: web/src/components/nomination/NominateForm.vue [hunk 9]
```diff
@@ -251,6 +253,7 @@ export default {
       if (!this.$v.form.speaker_bio.$dirty) { return errors; }
       if (!this.$v.form.speaker_bio.required) { errors.push(this.requiredError(this.$t('nominateSpeaker.bio.label'))); }
       if (!this.$v.form.speaker_bio.minLength) { errors.push(this.$t('validations.bioLength', [BIO_LENGTH])); }
+      if (!this.$v.form.speaker_bio.isThirdPerson) { errors.push(this.$t('validations.bioThirdPerson')); }
```
Comment (ann-kilzer) [inline web/src/components/nomination/NominateForm.vue:256]:
```suggestion
      if (this.$v.form.speaker_bio.isFirstPerson) { errors.push(this.$t('validations.bioThirdPerson')); }
```

File: web/tests/unit/validator-third-person.spec.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  it('accepts "I" if not first person', () => {
+    expect(isThirdPerson('She works in D&I')).to.be.true;
+  });
+  it('rejects first person in English', () => {
+    expect(isThirdPerson('I\' m a front end developer')).to.be.false;
+    expect(isThirdPerson('While in London, I studied things.')).to.be.false;
+  });
+  it('rejects first person in Japanese', () => {
+    expect(isThirdPerson('私はイタリア人です。')).to.be.false;
+    expect(isThirdPerson('わたし')).to.be.false;
```
Comment (ann-kilzer) [inline web/tests/unit/validator-third-person.spec.js:21]:
```suggestion　　
    expect(isFirstPerson('その建物は公私です！')).to.be.false;
    expect(isFirstPerson('公私は面白いです！')).to.be.false;
    expect(isFirstPerson('わたしは武士です！')).to.be.true;
```

**Review by ann-kilzer — COMMENTED** [review]
Super close, had another thought about naming now that I am seeing this from a real computer. Also made a suggestion to check for '^私は' or '^わたしは".  What do you think?

**Comment by tuttiq** [comment]
@RossellaFer @ann-kilzer I'm sorry for pointing this out late in the game, but I added my concerns about this feature [on the original ticket](https://github.com/WWCodeTokyo/speak-her-db/issues/119#issuecomment-701537380). Please let me know what you think.

**Comment by tuttiq** [comment]
@RossellaFer is it okay to close this PR? 🙇

**Comment by RossellaFer** [comment]
@tuttiq yes of course, I will close it now

**Pull request closed without merging**